### PR TITLE
Create report mutation

### PR DESCRIPTION
**Pull request by mtammineni**

### CODE REVIEW

**Review by mkrystof — DISMISSED** [review]

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/reports/ReportInputSchema.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    val jurisdiction: String? = null,
+
+    @GraphQLDescription("Sender ID this report belongs to; set to null if not applicable")
+    val senderId: String? = null,
+
+    @GraphQLDescription("Data Producer ID stated in the report; set to null if not applicable")
+    val dataProducerId: String? = null,
+
+    @GraphQLDescription("Content of the report. If the report is JSON then the content will be shown as JSON. Otherwise, the content is a base64 encoded string.")
+    val content: List<KeyValueInput>? = null,
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/reports/ReportInputSchema.kt]: Updated the "content" filed's datatype to be of the type, "String". Based on the "contentType" value, we will interpret and process the data in the "content" field.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/reports/ReportInputSchema.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
+package gov.cdc.ocio.processingstatusapi.models.reports
+
+import com.expediagroup.graphql.generator.annotations.GraphQLDescription
+import gov.cdc.ocio.processingstatusapi.models.submission.Aggregation
+import gov.cdc.ocio.processingstatusapi.models.submission.Status
+import java.time.OffsetDateTime
+
+@GraphQLDescription("Input type for a key-value pair")
+data class KeyValueInput(
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/reports/ReportInputSchema.kt]: Added class headers as applicable.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        return try {
+            performUpsert(input, action)
+        } catch (e: Exception) {
+            logger.error("Error upserting report: ${e.message}", e)
+            null
+        }
+    }
+
+    private fun performUpsert(input: ReportInput, action: String): Report? {
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt:81]: Added @Throws annotation for all the functions as relevant.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+     * @return The updated or newly created Report, or null if the operation fails.
+     * @throws BadRequestException If the action is invalid or the ID is improperly provided.
+     */
+    @Throws(BadRequestException::class)
+    private fun performUpsert(input: ReportInput, action: String): Report? {
+        // Validate action parameter
+        val upsertAction = UpsertAction.fromString(action)
+
+        // Validate required fields for ReportInput
+        if (input.dataStreamId.isNullOrBlank() || input.dataStreamRoute.isNullOrBlank() ||
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt]: Moved the validation section into validateInput function

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            UpsertAction.CREATE -> {
+                if (!report.id.isNullOrBlank()) {
+                    throw BadRequestException("ID should not be provided for create action.")
+                }
+
+                // Generate a new ID if not provided
+                report.id = generateNewId()
+                val options = CosmosItemRequestOptions()
+                val createResponse: CosmosItemResponse<Report>? = reportsContainer?.createItem(report, PartitionKey(report.uploadId!!), options)
+                createResponse?.item
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt]: Logging and Throwing ContentException

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+                }
+
+                // Attempt to read the existing item
+                val readResponse: CosmosItemResponse<Report>? = reportsContainer?.readItem(report.id!!, PartitionKey(report.uploadId!!), Report::class.java)
+
+                if (readResponse != null) {
+                    // Replace the existing item
+                    val options = CosmosItemRequestOptions()
+                    val replaceResponse: CosmosItemResponse<Report>? = reportsContainer?.replaceItem(report, report.id!!, PartitionKey(report.uploadId!!), options)
+                    replaceResponse?.item
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt]: Logging and Throwing ContentException

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+                }
+            }
+            ReportContentType.BASE64 -> {
+                // Decode base64 content into a Map, if expected
+                val decodedBytes = Base64.getDecoder().decode(content)
+                val decodedString = String(decodedBytes)
+                // If the decoded base64 string is in JSON format, parse it
+                try {
+                    objectMapper.readValue<Map<String, Any>>(decodedString)
+                } catch (e: Exception) {
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt:257]: Added logging

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+                    throw ContentException("Invalid JSON format after base64 decode: ${e.message}")
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt:259]: Added logging

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        // Validate action parameter
+        val upsertAction = UpsertAction.fromString(action)
+
+        // Validate required fields for ReportInput
+        if (input.dataStreamId.isNullOrBlank() || input.dataStreamRoute.isNullOrBlank() ||
+            input.stageInfo?.service.isNullOrBlank() || input.stageInfo?.action.isNullOrBlank()) {
+            throw BadRequestException("Missing required fields: dataStreamId, dataStreamRoute, stageInfo.service, and stageInfo.action must be present.")
+        }
+
+        val report = mapInputToReport(input)
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt]: Handled appropriately

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+        val validContentType = contentType?.let { ReportContentType.fromString(it) }
+
+        return when (validContentType) {
+            ReportContentType.JSON, ReportContentType.JSON_SHORT -> {
+                // Parse JSON content into a Map
+                try {
+                    objectMapper.readValue<Map<String, Any>>(content)
+                } catch (e: Exception) {
+                    throw ContentException("Invalid JSON format: ${e.message}")
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt:248]: Added logging

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutationService.kt [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+     * This function serves as a GraphQL mutation to create a new report or replace an existing one.
+     * It delegates the actual upsert logic to the ReportMutation class.
+     *
+     * @param input The ReportInput containing details of the report to be created or replaced.
+     * @param action A string specifying the action to perform: "create" or "replace".
+     * @return The result of the upsert operation, handled by the ReportMutation class.
+     */
+    @GraphQLDescription("Create upload")
+    @Suppress("unused")
+    @Throws(BadRequestException::class)
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutationService.kt:36]: Encapsulated inside try catch

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/reports/ReportInputSchema.kt [hunk 12]
```diff
@@ -0,0 +1,143 @@
+package gov.cdc.ocio.processingstatusapi.models.reports
+
+import com.expediagroup.graphql.generator.annotations.GraphQLDescription
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/models/reports/ReportInputSchema.kt]: Refactored into their own classes

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+    /**
+     * Maps the given ReportInput to a Report object.
+     *
+     * This method extracts the necessary fields from the input and constructs a Report instance.
+     * It also parses the content based on its type.
+     *
+     * @param input The ReportInput containing the details to map to a Report.
+     * @return A Report object populated with data from the input.
+     */
+    private fun mapInputToReport(input: ReportInput): Report {
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt:127]: Input cannot be null

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+                val decodedString = String(decodedBytes)
+                // If the decoded base64 string is in JSON format, parse it
+                try {
+                    objectMapper.readValue<Map<String, Any>>(decodedString)
+                } catch (e: Exception) {
+                    throw ContentException("Invalid JSON format after base64 decode: ${e.message}")
+                }
+            }
+            else -> {
+                throw ContentException("Unsupported content type: $contentType")
```
[the remark this answers is not in the document]
Comment (mtammineni) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/mutations/ReportMutation.kt:264]: Added logging

**Review by manu-govind — APPROVED** [review]

**Review by mkrystof — APPROVED** [review]

**Comment by mtammineni** [comment]
Create/ Replace Report feature added

**Pull request merged**